### PR TITLE
feat(integrations): add support for cluster clients from redis sdk

### PR DESCRIPTION
**Pull request by md384**
This change adds support for cluster clients from the redis sdk (as opposed to the rediscluster library).

This has also been tested in my own app which uses clusters (but not asyncio clusters).

Fixes GH-2523

### CODE REVIEW

**Comment by md384** [comment]
I have updated this PR with the db data on the span changes. For redis clusters, I have used the default_node but it might be better to not provide that data at all.

From redis docs - https://redis.readthedocs.io/en/stable/clustering.html#creating-clusters

> RedisCluster instance can be directly used to execute Redis commands. When a command is being executed through the cluster instance, the target node(s) will be internally determined. When using a key-based command, the target node will be the node that holds the key’s slot. Cluster management commands and other commands that are not key-based have a parameter called ‘target_nodes’ where you can specify which nodes to execute the command on. In the absence of target_nodes, the command will be executed on the default cluster node. As part of cluster instance initialization, the cluster’s default node is randomly selected from the cluster’s primaries, and will be updated upon reinitialization. Using r.get_default_node(), you can get the cluster’s default node, or you can change it using the ‘set_default_node’ method.

This likely means that it is difficult to get the actual node information and the default node will not always be the one used.

**Comment by wgiddens** [comment]
fixes #2523

**Review by szokeasaurusrex — COMMENTED** [review]
Thank you for your contribution!

I took an initial pass at your PR, and I added a few comments and a question. Please address these, and once I hear back, I will review again

**Comment by szokeasaurusrex** [comment]
@md384, to get your PR merged as quickly as possible, we would appreciate if you could [enable edits from repository maintainers](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests) if you are okay with doing so. This way, we can help with any minor changes that need to be made before merging your PR (such as resolving merge conflicts). Thanks!

File: tests/integrations/redis/asyncio/test_redis_asyncio.py
Comment (md384) [reply tests/integrations/redis/asyncio/test_redis_asyncio.py:15]: My mistake, this change is not needed. I will revert.

**Comment by md384** [comment]
> @md384, to get your PR merged as quickly as possible, we would appreciate if you could [enable edits from repository maintainers](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests) if you are okay with doing so. This way, we can help with any minor changes that need to be made before merging your PR (such as resolving merge conflicts). Thanks!

@szokeasaurusrex I have made your suggested changes and enabled access for maintainers. Thanks!

**Comment by szokeasaurusrex** [comment]
Thank you! I will take a look at your PR again when I have time

**Review by szokeasaurusrex — COMMENTED** [review]
Hey, thank you for addressing my previous suggestions. I have added a few more comments now; once you address them, I think this PR will be ready to merge!

File: sentry_sdk/integrations/redis/__init__.py
Comment (md384) [reply sentry_sdk/integrations/redis/__init__.py:153]: It does not, there is not much consistency across the redis modules

**Review by szokeasaurusrex — COMMENTED** [review]
Thank you for addressing the previous comments. I just tried to resolve the merge conflict between your branch and master, so your PR would be ready to merge, but I was unsure how to resolve it (please see my inline comment).

File: sentry_sdk/integrations/redis/__init__.py
Comment (md384) [reply sentry_sdk/integrations/redis/__init__.py:223]: I think we will need it for the `_set_db_data` but the other `set_db_data_fn`s will not - they check existence or use attributes that are required to have values. I will  change`_set_db_data`.

**Comment by szokeasaurusrex** [comment]
Hi @md384, now that I have resolved the merge conflict between your branch and master, I was able to run our CI checks on your code. It looks like your code is failing our linter ("CI / Lint Sources") and our Redis tests. The aws_lambda test is also failing for now, but this is expected, since I need to allow it to run separately; once you have fixed all the other tests, I will run the aws_lambda test.

Please fix the "CI / Lint Sources" and Redis tests, and let me know if you need help with anything!

File: tests/integrations/redis/cluster/test_redis_cluster.py
Comment (md384) [inline tests/integrations/redis/cluster/test_redis_cluster.py:38]: As I mention in the comment, initializing the RedisCluster executes a COMMAND call to the redis server to get available commands. This only happens on the first test run of a parameterized run (more of a problem for test_rediscluster_basic), perhaps there is some caching mechanism somewhere

File: sentry_sdk/integrations/redis/__init__.py
Comment (szokeasaurusrex) [inline sentry_sdk/integrations/redis/__init__.py:173]: Is there a way we could try to avoid using `_client` here, and instead access what we need through a public API? I would strongly prefer to avoid using an external library's private API, since the private API could change at any time and cause problems in the future.

File: tests/integrations/redis/cluster/test_redis_cluster.py
Comment (szokeasaurusrex) [reply tests/integrations/redis/cluster/test_redis_cluster.py:38]: I would rewrite the assertions as follows:
```suggestion
    # on initializing a RedisCluster, a COMMAND call is made - this is not important for the test
    # but must be accounted for
    assert len(crumbs) in (1, 2)
    assert len(crumbs) == 1 || crumbs[0]["message"] == "COMMAND"
```

The first assertion should be changed because the test should explicitly verify that at least one breadcrumb was created; the previous assertion would have succeeded with no breadcrumbs and the test would only fail later with an IndexError.

Regarding the second assertion, I think my suggested revision conveys more clearly the intent to verify that either there is only one breadcrumb, or if there are two, then the first breadcrumb should be a command.

If you prefer the structure with the `if` statement for the second assertion, it is okay with me to keep that structure, but in either case, please replace `event["breadcrumbs"]["values"][0]["message"]` with `crumbs[0]["message"]`

File: tests/integrations/redis/cluster/test_redis_cluster.py
Comment (szokeasaurusrex) [inline tests/integrations/redis/cluster/test_redis_cluster.py:40]:
```suggestion
    crumb = crumbs[-1]
```

File: tests/integrations/redis/cluster/test_redis_cluster.py
Comment (szokeasaurusrex) [inline tests/integrations/redis/cluster/test_redis_cluster.py:82]: My previous comment applies here, as well

**Review by szokeasaurusrex — COMMENTED** [review]
I have a few comments on your most recent changes, but I think this is getting really close

File: sentry_sdk/integrations/redis/__init__.py
Comment (szokeasaurusrex) [reply sentry_sdk/integrations/redis/__init__.py]: Okay, in that case, we can use `capture_internal_exceptions` here. I also checked that one of your unit tests is verifying this function, so we will get a test failure if `_client` ever gets removed

**Review by szokeasaurusrex — APPROVED** [review]
Looks good! Thanks for the contribution!

**Pull request merged**
